perception/autoware_lidar_frnet/lib/preprocess/preprocess_kernel.cu

```diff
@@ -0,0 +1,239 @@
+{
+__constant__ utils::FieldOfView const_fov;
+__constant__ utils::Dims2d const_interpolation;
+__constant__ utils::Dims2d const_frustum;
+
+PreprocessCuda::PreprocessCuda(const utils::PreprocessingParams & params, cudaStream_t stream)
+: interpolation_(params.interpolation), stream_(stream)
+{
+  CHECK_CUDA_ERROR(cudaMemcpyToSymbol(
+    const_fov, &params.fov, sizeof(utils::FieldOfView), 0, cudaMemcpyHostToDevice));
+  CHECK_CUDA_ERROR(cudaMemcpyToSymbol(
+    const_interpolation, &params.interpolation, sizeof(utils::Dims2d), 0, cudaMemcpyHostToDevice));
+  CHECK_CUDA_ERROR(cudaMemcpyToSymbol(
+    const_frustum, &params.frustum, sizeof(utils::Dims2d), 0, cudaMemcpyHostToDevice));
+}
+
+__device__ int2 project2d(const float3 point, const uint32_t scale_x, const uint32_t scale_y)
+{
+  // Compute the depth
+  float depth = sqrtf(point.x * point.x + point.y * point.y + point.z * point.z);
+
+  // Get angles (yaw and pitch)
+  float yaw = -atan2f(point.y, point.x);
+  float pitch = asinf(point.z / depth);
+
+  // Project to image coordinates (normalized)
+  float proj_x = 0.5f * (yaw / M_PIf + 1.0f);
+  float proj_y = 1.0f - (pitch + fabs(const_fov.down)) / const_fov.total;
+
+  // Scale to image size using angular resolution
+  proj_x *= static_cast<float>(scale_x);
+  proj_y *= static_cast<float>(scale_y);
+
+  // Return clamped and floor projection
+  return make_int2(
+    static_cast<int32_t>(fminf(fmaxf(proj_x, 0.0f), static_cast<float>(scale_x) - 1.0f)),
+    static_cast<int32_t>(fminf(fmaxf(proj_y, 0.0f), static_cast<float>(scale_y) - 1.0f)));
+}
+
+__global__ void projectPoints_kernel(
+  const InputPointType * cloud, const uint32_t num_points, uint32_t * output_num_points,
+  float * output_points, int64_t * output_coors, int64_t * output_coors_keys,
+  uint32_t * output_proj_idxs, float * output_proj_2d)
+{
+  uint32_t point_idx = blockIdx.x * blockDim.x + threadIdx.x;
+  if (point_idx >= num_points) return;
+
+  const InputPointType & point = cloud[point_idx];
+  const auto point3d = make_float3(point.x, point.y, point.z);
+  const auto proj_point = project2d(point3d, const_interpolation.w, const_interpolation.h);
+  const auto proj_coor = project2d(point3d, const_frustum.w, const_frustum.h);
+  const auto proj_idx = proj_point.y * const_interpolation.w + proj_point.x;
+  float4 & proj_2d_point = reinterpret_cast<float4 *>(output_proj_2d)[proj_idx];
+
+  // Writing to output arrays
+  atomicAdd(output_num_points, 1);
+
+  output_points[point_idx * 4 + 0] = point.x;
+  output_points[point_idx * 4 + 1] = point.y;
+  output_points[point_idx * 4 + 2] = point.z;
+  output_points[point_idx * 4 + 3] = point.intensity;
+
+  output_coors[point_idx * 3 + 0] = 0;
+  output_coors[point_idx * 3 + 1] = proj_coor.y;
+  output_coors[point_idx * 3 + 2] = proj_coor.x;
+
+  output_coors_keys[point_idx] = proj_coor.y * const_frustum.w + proj_coor.x;
+
+  // Update projection if not yet filled
+  if (!atomicExch(&output_proj_idxs[proj_idx], 1)) {
```

Comment:
AFAIK, float4 writes are not atomic, and most atomic operations operate up to 64 bits (blackwell can do 128 it seems), i.e., potential datarace
If you find a better solution lemme know but somewhere else I packed [data_x, range_32], [data_y, range_32], [data_z, range_32] and performed atomics